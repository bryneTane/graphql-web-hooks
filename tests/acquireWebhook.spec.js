const request = require('supertest')

const {
  migrateDatabase,
  seedDatabase,
  getNewServer,
  models,
  closeEverything,
  deleteTables,
  resetDatabase,
} = require('./test-database.js')
const {
  getNewClient,
  callWebhook
} = require('../lib/index')

// This is the maximum amount of time the band of test can run before timing-out
jest.setTimeout(600000)

let server = null
const client = getNewClient(
  `http://localhost:${process.env.PORT || 8080}/graphql`
)

const wait = (time) => new Promise((resolve) => setTimeout(resolve, time))

const acquireWebhook = (variables) => ({
  query: `mutation acquireWebhook($eventSecurityContext: JSON!, $eventType: String!) {
    acquireWebhook(eventSecurityContext: $eventSecurityContext, eventType: $eventType) {
      url
      headers {
        key
        value
      }
    }
  }`,
  variables,
  operationName: null,
})


/**
 * Starting the tests
 */
describe('Test acquireWebhook endpoint', () => {
  beforeAll(async () => {
    await migrateDatabase()
    await seedDatabase()
    server = await getNewServer()
  })

  beforeEach(async () => {
    await resetDatabase()
  })

  afterEach(async () => {
    await deleteTables()
  })

  afterAll(async () => {
    await closeEverything(server, models)
  })

  it('Can fetch webhook', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(
        acquireWebhook({
          eventSecurityContext: { userId: 1 },
          eventType: 'publish'
        })
      )

    expect(response.body.data).toMatchSnapshot()
  })

  it('Cannot fetch inexistent webhook', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(
        acquireWebhook({
          eventSecurityContext: { security: 'value' },
          eventType: 'publish'
        })
      )

    expect(response.body.errors).toMatchSnapshot()
  })

  it('Cannot fetch a webhook if event is not subscribed to it', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(
        acquireWebhook({
          eventSecurityContext: { userId: 1 },
          eventType: 'action'
        })
      )

    expect(response.body.errors).toMatchSnapshot()
  })

  it('Cannot fetch a webhook if event is not allowed', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(
        acquireWebhook({
          eventSecurityContext: { userId: 2 },
          eventType: 'publish'
        })
      )

    expect(response.body.errors).toMatchSnapshot()
  })
})
