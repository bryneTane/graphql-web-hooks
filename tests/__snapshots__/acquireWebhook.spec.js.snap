// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test acquireWebhook endpoint Can fetch webhook 1`] = `
Object {
  "acquireWebhook": Object {
    "headers": Array [
      Object {
        "key": "header-key-1",
        "value": "header-value-1",
      },
      Object {
        "key": "header-key-2",
        "value": "header-value-2",
      },
    ],
    "url": "https://www.google.com",
  },
}
`;

exports[`Test acquireWebhook endpoint Cannot fetch a webhook if event is not allowed 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "stacktrace": Array [
          "Error: Event is not allowed for this webhook",
          "    at Object.call (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:77:35)",
          "    at step (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:33:23)",
          "    at Object.next (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:14:53)",
          "    at fulfilled (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:5:58)",
        ],
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 2,
      },
    ],
    "message": "Event is not allowed for this webhook",
    "path": Array [
      "acquireWebhook",
    ],
  },
]
`;

exports[`Test acquireWebhook endpoint Cannot fetch a webhook if event is not subscribed to it 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "stacktrace": Array [
          "Error: This webhook is not subscribe at this event",
          "    at Object.call (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:73:39)",
          "    at step (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:33:23)",
          "    at Object.next (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:14:53)",
          "    at fulfilled (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:5:58)",
        ],
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 2,
      },
    ],
    "message": "This webhook is not subscribe at this event",
    "path": Array [
      "acquireWebhook",
    ],
  },
]
`;

exports[`Test acquireWebhook endpoint Cannot fetch inexistent webhook 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "stacktrace": Array [
          "Error: No webhook found",
          "    at Object.call (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:69:35)",
          "    at step (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:33:23)",
          "    at Object.next (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:14:53)",
          "    at fulfilled (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:5:58)",
        ],
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 2,
      },
    ],
    "message": "No webhook found",
    "path": Array [
      "acquireWebhook",
    ],
  },
]
`;

exports[`Test call webhook Can fetch webhook 1`] = `
Object {
  "acquireWebhook": Object {
    "headers": Array [
      Object {
        "key": "header-key-1",
        "value": "header-value-1",
      },
      Object {
        "key": "header-key-2",
        "value": "header-value-2",
      },
    ],
    "url": "https://www.google.com",
  },
}
`;

exports[`Test call webhook Cannot fetch a webhook if event is not allowed 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "stacktrace": Array [
          "Error: Event is not allowed for this webhook",
          "    at Object.call (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:77:35)",
          "    at step (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:33:23)",
          "    at Object.next (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:14:53)",
          "    at fulfilled (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:5:58)",
        ],
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 2,
      },
    ],
    "message": "Event is not allowed for this webhook",
    "path": Array [
      "acquireWebhook",
    ],
  },
]
`;

exports[`Test call webhook Cannot fetch a webhook if event is not subscribed to it 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "stacktrace": Array [
          "Error: This webhook is not subscribe at this event",
          "    at Object.call (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:73:39)",
          "    at step (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:33:23)",
          "    at Object.next (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:14:53)",
          "    at fulfilled (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:5:58)",
        ],
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 2,
      },
    ],
    "message": "This webhook is not subscribe at this event",
    "path": Array [
      "acquireWebhook",
    ],
  },
]
`;

exports[`Test call webhook Cannot fetch inexistent webhook 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "stacktrace": Array [
          "Error: No webhook found",
          "    at Object.call (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:69:35)",
          "    at step (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:33:23)",
          "    at Object.next (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:14:53)",
          "    at fulfilled (/Users/victorgermain/Documents/work/graphql-web-hook/lib/graphql/webhook/acquire.js:5:58)",
        ],
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 2,
      },
    ],
    "message": "No webhook found",
    "path": Array [
      "acquireWebhook",
    ],
  },
]
`;
